apps/io_tester/io_tester.cc

```diff
@@ -657,6 +665,99 @@ class write_io_class_data : public io_class_data {
+            throw std::runtime_error("request_type::unlink requires specifying 'files_count'");
+        }
+    }
+
+    future<> do_start(sstring path, directory_entry_type type) override {
+        if (type == directory_entry_type::directory) {
+            return do_start_on_directory(path);
+        }
+        throw std::runtime_error(format("Unsupported storage. {} should be directory", path));
+    }
+
+    future<size_t> issue_request(char *buf, io_intent* intent) override {
+        if (all_files_removed()) {
+            return make_ready_future<size_t>(1u);
+        }
+
+        const auto fname = get_filename(_file_id_to_remove);
+        ++_file_id_to_remove;
+
+        return remove_file(fname).then([]{
+            return make_ready_future<size_t>(1u);
+        });
+    }
+
+    void emit_results(YAML::Emitter& out) override {
+        out << YAML::Key << "average latency" << YAML::Value << average_latency();
+    }
+
+private:
+    uint64_t files_count() const {
+        return *_config.files_count;
+    }
+
+    bool all_files_removed() const {
+        return files_count() <= _file_id_to_remove;
+    }
+
+    sstring get_filename(uint64_t file_id) const {
+        return format("{}/test-{}-shard-{:d}-file-{}", _dir_path, name(), this_shard_id(), file_id);
+    }
+
+    future<> do_start_on_directory(sstring path) {
+        _dir_path = std::move(path);
+
+        // When we have many files it is easy to exceed the limit of open file descriptors.
+        // To avoid that the limit is divided between shards (leaving some room for other jobs).
+        const uint64_t max_concurrency = static_cast<uint64_t>((1024u / smp::count) * 0.8);
+        return max_concurrent_for_each(boost::irange(UINT64_C(0), files_count()), max_concurrency, [this] (uint64_t file_id) {
+            const auto fname = get_filename(file_id);
+            const auto fsize = _config.file_size / files_count();
+            const auto flags = open_flags::rw | open_flags::create;
+
+            file_open_options options;
+            options.extent_allocation_size_hint = fsize;
+            options.append_is_unlikely = true;
+
+            return open_file_dma(fname, flags, options).then([fsize] (auto f) mutable {
+                return do_with(std::move(f), [fsize] (auto f) {
+                    return f.size().then([f, fsize] (uint64_t pre_truncate_size) mutable {
+                        return f.truncate(fsize).then([f, fsize, pre_truncate_size] () {
+                            if (pre_truncate_size >= fsize) {
+                                return make_ready_future<>();
+                            }
+
+                            const uint64_t buffer_size{256ul << 10};
+                            const uint64_t buffers_count{static_cast<uint64_t>(fsize / buffer_size) + 1u};
+
+                            return do_with(boost::irange(UINT64_C(0), buffers_count), [f, buffer_size] (auto& buffers_range) mutable {
```

Comment:
Please, generalize this code with io_class_data::do_start_on_directory()

Reply:
Sure. Extracted the common part to a free function called `create_and_fill_file()`.

This function is invoked by `io_class_data` and `unlink_class_data` during `do_start_on_directory()`. It returns the pair: handle to file and the offset to the last write position - the latter is needed in case of `request_type::append`.